### PR TITLE
mtmd: use align_corners for Qwen3-VL vision position-embedding interpolation

### PR DESCRIPTION
**Pull request by gerbenvv**
## Summary

Qwen3-VL's learned absolute vision position embedding is interpolated to the runtime patch grid in
`tools/mtmd/models/qwen3vl.cpp` via `resize_position_embeddings()`, which uses the default
`GGML_SCALE_MODE_BILINEAR | GGML_SCALE_FLAG_ANTIALIAS` - in other words, **align_corners=False** (half-pixel).

However, the reference transformers implementation uses **align_corners=True**.
`transformers` its `get_vision_bilinear_indices_and_weights` builds the sampling grid with
`torch.linspace(0, side - 1, T)`, so pinned to `[0, side-1]`.

This fixes that.

## Additional information

This is regarding my own experience and the comment at the end of #16880 (which is now closed).
The 0–1000 relative-coordinate rescale discussed is correct, but @sucubus666's point was also correct:
after applying the 0–1000 rescale there is still a residual spatial error that grows with image size,
is per-axis for non-square images, and isn't corrected for. That residual is this backend
interpolation-convention mismatch.

You can verify this yourself using a non-square image and prompting it to do detection of something
small, like a word. It will output a box that appears to be scaled outwards from the center of the
image.

## Requirements

- I have read and agree with the [contributing guidelines](https://github.com/ggml-org/llama.cpp/blob/master/CONTRIBUTING.md): YES
- AI usage disclosure: YES, it helped me find the problem. However, I made this PR and reviewed and compared it myself as well, making sure it's correct.

### CODE REVIEW

**Comment by gerbenvv** [comment]
Sorry about not following the format, fixed it just now and re-opened.

**Comment by gerbenvv** [comment]
Working now on a full independent reproduction that will prove without a doubt that this is the correct way :grin:

**Comment by gerbenvv** [comment]
Full reproduction with:

```python
import ctypes
import os
import re
from pathlib import Path

import llama_cpp as llama
import llama_cpp.mtmd_cpp as mtmd
import numpy as np
from huggingface_hub import hf_hub_download
from PIL import Image, ImageDraw, ImageFont

WIDTH: int = 1792
HEIGHT: int = 2560
WORD: str = "ELEPHANT"

# The grounding prompt; the model answers with `[x1, y1, x2, y2]` in [0, 1000].
PROMPT: str = (
    f'Detect the word "{WORD}" in the image and output its bounding box as [x1, y1, x2, y2] '
    "with integer coordinates normalized from 0 to 1000. Output only the box."
)

OUTPUT_FILE_PATH: Path = Path("output.png")


def make_image() -> tuple[Image.Image, tuple[int, int, int, int]]:
    """Renders the word on a white canvas, returning the image and its ground-truth box."""

    image = Image.new("RGB", (WIDTH, HEIGHT), "white")
    draw = ImageDraw.Draw(image)
    font = ImageFont.truetype("/usr/share/fonts/truetype/hack/Hack-Regular.ttf", 130)

    # Place the word near the lower-right corner, far from the image center where the
    # `align_corners` center-scaling error is largest.
    anchor = (int(0.42 * WIDTH), int(0.78 * HEIGHT))
    draw.text(anchor, WORD, fill="black", font=font)
    ground_truth = draw.textbbox(anchor, WORD, font=font)

    return image, ground_truth


def parse_box(text: str) -> tuple[float, float, float, float] | None:
    """Parses the last `[x1, y1, x2, y2]` group (in [0, 1000]) into pixel coordinates."""

    # Drop any `<think>...</think>` block, then take the last coordinate group.
    text = re.sub(r"<think>.*?</think>", "", text, flags=re.DOTALL)
    groups = re.findall(r"\[\s*(-?\d+)\s*,\s*(-?\d+)\s*,\s*(-?\d+)\s*,\s*(-?\d+)\s*\]", text)
    if not groups:
        return None

    x1, y1, x2, y2 = (int(value) for value in groups[-1])

    return x1 / 1000 * WIDTH, y1 / 1000 * HEIGHT, x2 / 1000 * WIDTH, y2 / 1000 * HEIGHT


def main() -> None:
    image, ground_truth = make_image()

    model_file_path = hf_hub_download("unsloth/Qwen3.5-0.8B-GGUF", "Qwen3.5-0.8B-BF16.gguf")
    mmproj_file_path = hf_hub_download("unsloth/Qwen3.5-0.8B-GGUF", "mmproj-BF16.gguf")

    # Load the text model.
    llama.llama_backend_init()
    model_params = llama.llama_model_default_params()
    model_params.n_gpu_layers = 999
    model = llama.llama_model_load_from_file(model_file_path.encode(), model_params)
    vocab = llama.llama_model_get_vocab(model)

    context_params = llama.llama_context_default_params()
    context_params.n_ctx = 8192
    context_params.n_batch = 2048
    context_params.n_ubatch = 2048
    context = llama.llama_init_from_model(model, context_params)

    # Load the vision projector. The token limits are set wide so llama.cpp does not smart-resize
    # the image, leaving the compiled `align_corners` of the position-embedding interpolation as
    # the only difference between the stock and patched wheels.
    mtmd_params = mtmd.mtmd_context_params_default()
    mtmd_params.use_gpu = True
    mtmd_params.n_threads = os.cpu_count() or 8
    mtmd_params.media_marker = mtmd.mtmd_default_marker()
    mtmd_params.image_min_tokens = 1
    mtmd_params.image_max_tokens = 1_000_000
    mtmd_context = mtmd.mtmd_init_from_file(mmproj_file_path.encode(), model, mtmd_params)

    # Wrap the image pixels in an mtmd bitmap.
    pixels = np.ascontiguousarray(np.asarray(image, dtype=np.uint8))
    pixel_buffer = (ctypes.c_uint8 * pixels.size).from_buffer_copy(pixels.tobytes())
    bitmap = mtmd.mtmd_bitmap_init(WIDTH, HEIGHT, pixel_buffer)

    # Build the prompt with an empty think block, so the model skips reasoning and emits the box.
    marker = mtmd.mtmd_default_marker().decode()
    prompt = (
        f"<|im_start|>user\n<|vision_start|>{marker}<|vision_end|>{PROMPT}<|im_end|>\n"
        f"<|im_start|>assistant\n<think>\n\n</think>\n\n"
    )

    input_text = mtmd.mtmd_input_text()
    input_text.text = prompt.encode()
    input_text.add_special = True
    input_text.parse_special = True

    chunks = mtmd.mtmd_input_chunks_init()
    bitmaps = (mtmd.mtmd_bitmap_p_ctypes * 1)(bitmap)
    if mtmd.mtmd_tokenize(mtmd_context, chunks, ctypes.byref(input_text), bitmaps, 1) != 0:
        raise RuntimeError("mtmd_tokenize failed.")

    print("image tokens:", mtmd.mtmd_helper_get_n_tokens(chunks))

    # Evaluate the image and prompt; the helper sets the M-RoPE positions.
    n_past = ctypes.c_int32(0)
    status = mtmd.mtmd_helper_eval_chunks(
        mtmd_context,
        context,
        chunks,
        ctypes.c_int32(0),
        ctypes.c_int32(0),
        ctypes.c_int32(2048),
        True,
        ctypes.byref(n_past),
    )
    if status != 0:
        raise RuntimeError("mtmd_helper_eval_chunks failed.")

    # Greedily decode the answer.
    num_vocab = llama.llama_vocab_n_tokens(vocab)
    tokens = []
    for _ in range(48):
        logits = np.ctypeslib.as_array(llama.llama_get_logits_ith(context, -1), shape=(num_vocab,))
        token = int(np.argmax(logits))
        if llama.llama_vocab_is_eog(vocab, token):
            break

        tokens.append(token)
        batch = llama.llama_batch_get_one((llama.llama_token * 1)(token), 1)
        if llama.llama_decode(context, batch) != 0:
            break

    pieces = []
    for token in tokens:
        piece_buffer = ctypes.create_string_buffer(64)
        length = llama.llama_token_to_piece(vocab, token, piece_buffer, len(piece_buffer), 0, True)
        pieces.append(piece_buffer.raw[:length].decode("utf-8", "replace"))
    text = "".join(pieces)

    box = parse_box(text)
    print("llama-cpp-python", llama.__version__)
    print("ground truth :", ground_truth)
    print("raw output   :", text.strip())
    print("box (px)     :", tuple(round(value) for value in box) if box else None)

    # Draw the ground-truth box and the predicted box, then save.
    annotated = image.copy()
    draw = ImageDraw.Draw(annotated)
    draw.rectangle(ground_truth, outline=(0, 0, 0), width=3)
    if box:
        label = f"llama.cpp box ({llama.__version__})"
        draw.rectangle(box, outline=(0, 90, 255), width=6)
        draw.text((box[0] + 4, max(0, box[1] - 16)), label, fill=(0, 90, 255))
    annotated.save(OUTPUT_FILE_PATH)
    print("saved:", OUTPUT_FILE_PATH)


if __name__ == "__main__":
    main()
```

Difference here is quite big in fact!

Before:

qwen3vl_before

After:
qwen3vl_after

**Review by xuanson2025 — APPROVED** [review]
Thanks for the fix. If possible, could you also check if this fix is applicable to other qwen models? i.e. qwen2 and qwen 2.5?

@ggerganov this is good to merge

**Comment by gerbenvv** [comment]
@xuanson2025 Will try today if the fix also applies to the other Qwen VL models.

**Comment by gerbenvv** [comment]
It looks like Qwen 2-VL/2.5-VL has no learned absolute position embedding to interpolate.
Their vision encoder positions come from just 2D rotary (M-RoPE), whereas Qwen 3-VL adds the learned absolute position embedding.

- qwen2vl.cpp: ggml_rope_multi(..., GGML_ROPE_TYPE_VISION, ...) with mrope_sections; no model.position_embeddings, no resize_position_embeddings, no ggml_interpolate
- qwen3vl.cpp: GGML_ASSERT(model.position_embeddings != nullptr) + resize_position_embeddings()

So the fix is only needed for Qwen 3-VL.

**Comment by gerbenvv** [comment]
@xuanson2025 @ggerganov This is good to merge now (since no other models affected).

**Review by ngxson — APPROVED** [review]
